app/backends/_common/js/worker.js

```diff
@@ -0,0 +1,158 @@
+                    inputs: [],
+                    expected: typeof hiddenPuzzle.expected === 'string' ? JSON.parse(hiddenPuzzle.expected) : hiddenPuzzle.expected,
+                };
+
+                counter = -1;
+                while (++counter < inputCopies) {
+                    hiddenContent.inputs.push(_.cloneDeep(hiddenContent.input));
+                }
+
+                content.hidden.push(hiddenContent);
+            }
+        }
+    }
+
+    payload.content = undefined; // no need to transfer back the filler
+
+    postMessage({
+        type: 'wrk::filled',
+        payload: payload
+    });
+}
+
+function evaluate(task, puzzleContent) {
+    try {
+        var result = eval('(function(_, content, global, setTimeout, setInterval, self, location, navigator, onmessage, close, postMessage, importScripts) {"use strict"; return (function (arg) {' + task.input + '})(content);}).call(stub, stub, puzzleContent, stub, stubFun, stubFun, stub, stub, stub, stubFun, stubFun, stubFun, stubFun)');
+
+        return [null, result];
+    } catch (e) {
+        return [e, null];
+    }
+}
+
+function evaluateHidden(task, hidden) {
+    var index = -1;
+    var hiddenCount = hidden.length;
+
+    while (++index < hiddenCount) {
+        var hiddenPuzzle = hidden[index];
+        var puzzleContent = hiddenPuzzle.inputs.pop() || _.cloneDeep(hiddenPuzzle.input);
+        var evalResult = evaluate(task, puzzleContent);
+
+        if (evalResult[0]) {
+            return false;
+        }
+
+        if (!isEqual(hiddenPuzzle.expected, evalResult[1])) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function exec(payload) {
+    var puzzleContent = content.inputs.pop() || _.cloneDeep(content.input);
+    var evalResult = evaluate(payload.task, puzzleContent);
+
+    if (evalResult[0]) {
+        // error branch
+        var e = evalResult[0];
+        payload.error = ((e && e.payload) || (e + ''));
+    } else {
+        var result = evalResult[1];
+        var resultStr = JSON.stringify(result);
+
+        if (typeof resultStr === 'undefined') {
+            resultStr = resultBackup; // undefined, function, Symbol are stringify-ed to undefined
+        }
+        payload.result = resultStr;
+
+        if (isEqual(content.expected, result)) {
+            if (evaluateHidden(payload.task, content.hidden)) {
+                payload.correct = 'correct';
+            } else {
+                payload.correct = 'partial';
+            }
+        } else {
+            payload.correct = 'incorrect';
+        }
+    }
+
+    postMessage({
+        type: 'wrk::done',
+        payload: payload
+    });
+}
+
+self.addEventListener('message', function (event) {
+    var data = event.data;
+
+    switch (data.type) {
+        case 'wrk:>init':
+            return initWorker(data.payload);
+        case 'wrk:>fill':
+            return fillWorker(data.payload);
+        case 'wrk:>exec':
```

Comment:
just curious about this convention. What `::` and `:>` stands for?

Reply:
`:>` - message to worker (usually some command).
`::` - message from worker (reaction on command or notification about worker's state)